### PR TITLE
feat: when you create a new schema you can now choose a template M8564

### PR DESCRIPTION
**Pull request by mswertz**
Simplified: when you create a database/schema you can now choose a template, and whether you like to include demo data.

Side effect of pull request is also cleaning up the task api, and expanding its use.
Todo: also show task progress during schema creation. Therefore components have been created (can be added in later pull request)

Important note: also moves catalogue data model to another place so they can be loaded from classpath!!!

### CODE REVIEW

**Comment by connoratrug** [comment]
would suggest hiding the advanced options by default and using a radio ( small set) or select ui to select the template. The source url is seems confusing to me, especially as creating a db is maybe one of the fist things a user evaluating the project would do

File: apps/central/src/components/SchemaCreateModal.vue [hunk 1]
```diff
@@ -50,17 +50,22 @@
...
-          >
-            Cohort Catalogue
-          </ButtonAlt>
+          <ul>
+            <li>
+              <ButtonAlt
+                @click="
+                  schemaName = 'CohortCatalogue';
+                  sourceURL = [
+                    host + '/public_html/apps/data/datacatalogue',
```
Comment (connoratrug) [inline apps/central/src/components/SchemaCreateModal.vue:59]: this ties the frontend to specific backend details ( that should probably remain hidden ) , suggest passing the id of a template instead of more low level url

File: backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlDatabaseFieldFactory.java [hunk 2]
```diff
@@ -6,14 +6,14 @@
...
+import java.net.URL;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
-import org.molgenis.emx2.Constants;
-import org.molgenis.emx2.Database;
-import org.molgenis.emx2.SchemaInfo;
+import org.molgenis.emx2.*;
```
Comment (connoratrug) [inline backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlDatabaseFieldFactory.java:15]: suggest explicit imports

File: backend/molgenis-emx2-io/src/main/java/org/molgenis/emx2/io/tablestore/TableStoreForURL.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+/** We test this in webapi */
+public class TableStoreForURL implements TableAndFileStore {
+  private URL baseURL;
+
+  public TableStoreForURL(URL baseURL) {
+    if (!baseURL.toString().endsWith("/")) {
+      try {
+        this.baseURL = new URL(baseURL + "/");
+      } catch (Exception e) {
+        // should never happen
```
Comment (connoratrug) [inline backend/molgenis-emx2-io/src/main/java/org/molgenis/emx2/io/tablestore/TableStoreForURL.java:26]: and will never know if it does happen, suggest handeling the error

**Comment by mswertz** [comment]
> would suggest hiding the advanced options by default and using a radio ( small set) or select ui to select the template. The source url is seems confusing to me, especially as creating a db is maybe one of the fist things a user evaluating the project would do

I have mixed feelings here. You can use this sourceURl also for external users so don't want to hide the mechanism completely from the user. However, I also considered to hide it under 'advanced'. I wonder what data managers say, because this screen is 'admin' so users should be rather advanced anyway.

File: backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlDatabaseFieldFactory.java (hunk 2 above)
Comment (mswertz) [reply backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlDatabaseFieldFactory.java:15]: this is done automatically by our stylechecker

File: backend/molgenis-emx2-io/src/main/java/org/molgenis/emx2/io/tablestore/TableStoreForURL.java (hunk 3 above)
Comment (mswertz) [reply backend/molgenis-emx2-io/src/main/java/org/molgenis/emx2/io/tablestore/TableStoreForURL.java:26]: Sure, I will make a useless throw out of it.

File: apps/central/src/components/SchemaCreateModal.vue (hunk 1 above)
Comment (mswertz) [reply apps/central/src/components/SchemaCreateModal.vue:59]: I considered this then then rejected this idea. I think it is useful for data managers to understand what is going on here, so they could use this also for other data models.

**Comment by erikzwart** [comment]
Error while trying to load cohort template:
image

The option load from source will not complain if no zip is found and creates empty schema, unclear to me which url to use:
image

**Comment by mswertz** [comment]
[#8564 As catalogue SAAS developer i want reference schema implementation as part of the code base and deployment](https://trello.com/c/O2auMx03/397-8564-as-catalogue-saas-developer-i-want-reference-schema-implementation-as-part-of-the-code-base-and-deployment)

**Comment by mswertz** [comment]
We decided to redesign as a 'setup wizard' where you can either create an empty database it is now, or you can install an existing database model. This in wizard also allows for some questions, such as choosing to include/not include data.

**Comment by mswertz** [comment]
simplified user experience.
And also simplified engineering dropping the whole sourceURL thing.
Unfortunately huge pull request now.

**Comment by mswertz** [comment]
Updated, see initial comment.

File: apps/central/src/components/SchemaCreateModal.vue [hunk 4]
```diff
@@ -98,6 +119,11 @@ export default {
     endpoint() {
       return "/api/graphql";
     },
+    host() {
+      if (window) {
+        return window.location.origin;
+      }
```
Comment (connoratrug) [inline apps/central/src/components/SchemaCreateModal.vue:125]: else ?

File: apps/graphql-playground/src/index.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import { createGraphiQLFetcher } from "@graphiql/toolkit";
+
+const fetcher = createGraphiQLFetcher({
+  url: window.location.origin + "/graphql",
+});
+
+ReactDOM.render(
+  <GraphiQL fetcher={fetcher} editorTheme={"dracula"} />,
+  document.body
+);
```
Comment (connoratrug) [inline apps/graphql-playground/src/index.js:14]: why is this added ?

Comment (mswertz) [reply apps/graphql-playground/src/index.js:14]: thx. this is an error, will remove it.

File: apps/central/src/components/SchemaCreateModal.vue (hunk 4 above)
Comment (mswertz) [reply apps/central/src/components/SchemaCreateModal.vue:125]: this was added in light of server side rendering. Else it will not work :-) Do you have recommendation?

**Comment by mswertz** [comment]
Obviously I will fix build.

File: apps/central/src/components/SchemaCreateModal.vue (hunk 4 above)
Comment (connoratrug) [reply apps/central/src/components/SchemaCreateModal.vue:125]: in case of ssr host will return null, but i need to look in to the use of the host var here. For now i do not see a problem as the central add does not run via ssr

File: backend/molgenis-emx2/src/main/resources/log4j2.xml [hunk 6]
```diff
@@ -7,9 +7,9 @@
     </Appenders>
     <Loggers>
         <Logger name="org.pac4j" level="debug" additivity="false">
-            <Appender-ref ref="Console" level="debug" />
+            <Appender-ref ref="Console" level="debug"/>
         </Logger>
-        <Root level="error">
```
Comment (connoratrug) [inline backend/molgenis-emx2/src/main/resources/log4j2.xml:12]: by accident ? or do you want to change the default ( why and maybe move to separate pr)

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/GraphqlApi.java [hunk 7]
```diff
@@ -66,8 +66,8 @@ public static String handleSchemaRequests(Request request, Response response) th
     MolgenisSession session = sessionManager.getSession(request);
     String schemaName = sanitize(request.params(SCHEMA));
 
-    // apps is not a schema but a resource
-    if ("apps".equals(schemaName)) {
+    // apps and api is not a schema but a resource
+    if ("apps".equals(schemaName) || "api".equals(schemaName)) {
```
Comment (connoratrug) [inline backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/GraphqlApi.java:70]: could be solved by defining the routes in the 'correct' order , this feels a bit hacky

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/RunMolgenisEmx2.java [hunk 8]
```diff
@@ -33,8 +33,7 @@ public static void main(String[] args) {
       db.becomeAdmin();
       if (db.getSchema("pet store") == null) {
         Schema schema = db.createSchema("pet store");
-        PetStoreExample.create(schema.getMetadata());
-        PetStoreExample.populate(schema);
+        new PetStoreLoader().load(schema, true);
```
Comment (connoratrug) [inline backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/RunMolgenisEmx2.java:36]: what does true do , maybe use variable name to make clear what the property is meant for

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/TaskServiceInMemory.java [hunk 9]
```diff
@@ -44,12 +40,17 @@ public Task getTask(String id) {
...
   }
 
+  @Override
+  public Collection<Task> listTasks() {
+    return tasks.values();
+  }
+
   @Override
   public void removeOlderThan(long milliseconds) {
     Set<String> keys = tasks.keySet();
```
Comment (connoratrug) [inline backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/TaskServiceInMemory.java:50]: forEach ?

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * For documenting processes that consist of multiple steps and elements. For example. batch
+ * insert/upload tasks.
+ */
+public class Task implements Runnable, Iterable<Task> {
+  // some unique id
+  private String id = UUID.randomUUID().toString();
```
Comment (connoratrug) [inline backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java:19]: final ?

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+  // human readable description
+  private String description;
+  // status of the tas
+  private TaskStatus status = WAITING;
+  // total size of the task, used for progress monitoring
+  private Integer total;
+  // position in the total size, if available, used for progress monitoring
+  private Integer progress;
+  // start time to measure run time
+  private long startTimeMilliseconds = System.currentTimeMillis();
```
Comment (connoratrug) [inline backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java:33]: final

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+    return step;
+  }
+
+  public void addSubTask(Task task) {
+    Objects.requireNonNull(task, "task cannot be null");
+    this.subTasks.add(task);
+  }
+
+  public List<Task> getSubTasks() {
+    return subTasks;
```
Comment (connoratrug) [inline backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java:76]: this.subTasks ? , probably want to return immutable list

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+  }
+
+  public Integer getProgress() {
+    if (RUNNING.equals(status)) return progress;
```
Comment (connoratrug) [inline backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java:80]: this.status ? , suggest to alway return the return type ( instead of null ) , should add a note about the range of the returnd int ( between 0 and 100 ? )

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+    return null;
+  }
+
+  public Task setProgress(int progress) {
```
Comment (connoratrug) [inline backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java:84]: is it oke to call with progress = 100000 ?

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+    this.status = status;
+    return this;
+  }
+
+  public void skipped(String description) {
+    this.skipped();
+    this.setDescription(description);
+  }
+
+  public void skipped() {
```
Comment (connoratrug) [inline backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java:185]: setSkipped or some thing to indicate its a setter

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+    this.setStatus(ERROR);
+  }
+
+  public void error(String description) {
+    this.error();
+    this.setDescription(description);
+  }
+
+  @Override
+  public void run() {}
```
Comment (connoratrug) [inline backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java:201]: why override with empty impl

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+
+  public boolean isStrict() {
+    return strict;
+  }
+
+  public String toString() {
+    try {
+      return mapper.writeValueAsString(this);
+    } catch (Exception e) {
+      // should never happen
```
Comment (connoratrug) [inline backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java:226]: this is case for all exceptions

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+import java.util.function.Consumer;
+import org.molgenis.emx2.MolgenisException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * For documenting processes that consist of multiple steps and elements. For example. batch
+ * insert/upload tasks.
+ */
+public class Task implements Runnable, Iterable<Task> {
```
Comment (connoratrug) [inline backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java:17]: missing equals anh hash code

File: backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlApiMutationResult.java [hunk 19]
```diff
@@ -21,12 +21,18 @@ public enum Status {
...
   public GraphqlApiMutationResult(Status status, String message, Object... formatValues) {
     this.status = status;
     this.message = String.format(message, formatValues);
   }
 
+  public GraphqlApiMutationResult setTaskId(String taskId) {
+    this.taskId = taskId;
+    return this;
+  }
+
```
Comment (connoratrug) [inline backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlApiMutationResult.java:35]: maybe use split into two separate mutation results ( with and without task id )

File: backend/molgenis-emx2-graphql/src/test/java/org/molgenis/emx2/graphql/TestGraphQLCompositeKeys.java [hunk 20]
```diff
@@ -28,7 +28,7 @@ public class TestGraphQLCompositeKeys {
   public static void setup() {
     database = TestDatabaseFactory.getTestDatabase();
     Schema schema = database.dropCreateSchema(schemaName);
-    grapql = new GraphqlApiFactory().createGraphqlForSchema(schema);
+    grapql = new GraphqlApiFactory().createGraphqlForSchema(schema, null);
```
Comment (connoratrug) [inline backend/molgenis-emx2-graphql/src/test/java/org/molgenis/emx2/graphql/TestGraphQLCompositeKeys.java:31]: maybe use method overload

File: backend/molgenis-emx2-graphql/src/test/resources/dataexample/molgenis.csv [hunk 21]
```diff
@@ -0,0 +1,3 @@
+tableName,columnName,key
```
Comment (connoratrug) [inline backend/molgenis-emx2-graphql/src/test/resources/dataexample/molgenis.csv:1]: check usage

File: backend/molgenis-emx2-graphql/src/test/resources/dataexample/test.csv [hunk 22]
```diff
@@ -0,0 +1,2 @@
+col1,col2,colerror
```
Comment (connoratrug) [inline backend/molgenis-emx2-graphql/src/test/resources/dataexample/test.csv:1]: check usage

File: backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlApiMutationResult.java (hunk 19 above)
Comment (mswertz) [reply backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlApiMutationResult.java:35]: solved on graphql level (which does have two different mutationresult)

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/TaskServiceInMemory.java (hunk 9 above)
Comment (mswertz) [reply backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/TaskServiceInMemory.java:50]: ?

**Comment by connoratrug** [comment]
When creating a schema from template and loading data the succes msg should probably not show the links to schema and upload, link to tables or just redirect to tables

**Comment by connoratrug** [comment]
suggest to a message to create schema spinner as in the case of a catalogue with data this can take more then a few seconds

File: apps/catalogue/logfile [hunk 23]
```diff
@@ -0,0 +1,2 @@
+2022-04-14 08:44:17.014 CEST [73861] FATAL:  lock file "postmaster.pid" already exists
+2022-04-14 08:44:17.014 CEST [73861] HINT:  Is another postmaster (PID 73851) running in data directory "/opt/homebrew/var/postgres"?
```
Comment (connoratrug) [inline apps/catalogue/logfile:2]: remove this file ?

File: docs/armadillo/_sidebar.md [hunk 24]
```diff
@@ -1,4 +0,0 @@
-- [MOLGENIS Armadillo guide](README.md)
```
Comment (connoratrug) [inline docs/armadillo/_sidebar.md:1]: other pr ?

File: docs/_sidebar.md [hunk 25]
```diff
@@ -1,4 +1,36 @@
-- [Start](/)
```
Comment (connoratrug) [inline docs/_sidebar.md:1]: other pr ?

File: backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/TaskServiceInMemory.java (hunk 9 above)
Comment (connoratrug) [reply backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/TaskServiceInMemory.java:50]: why get a key list and then do lots of lookups , instead of applying function to each item , the whole for the the task.get(index).prop makes it hard to read, could be tasks.forEach(doSomething), or in case of remove https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#removeIf-java.util.function.Predicate-

**Review by connoratrug — CHANGES_REQUESTED** [review]
seems like a few other changes got mixed in with the pr

**Comment by mswertz** [comment]
> seems like a few other changes got mixed in with the pr

removed

**Review by connoratrug — COMMENTED** [review]
i don't get it , now the template fields says required ( but its not) , and in the case you do not use the template the success msg do not show the links to editor and upload

**Review by connoratrug — APPROVED** [review]
